consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/payload/MessageFactory.java

```diff
@@ -34,19 +42,73 @@ public MessageFactory(final NodeKey nodeKey) {
     this.nodeKey = nodeKey;
   }
 
+  public Proposal createProposal(
+      final ConsensusRoundIdentifier roundIdentifier,
+      final Block block,
+      final Optional<RoundChangeCertificate> roundChangeCertificate) {
+
+    final ProposalPayload payload = new ProposalPayload(roundIdentifier, block);
+
+    return new Proposal(
+        createSignedMessage(payload),
+        roundChangeCertificate
+            .map(rc -> new ArrayList<>(rc.getRoundChangePayloads()))
+            .orElse(new ArrayList<>()),
+        roundChangeCertificate
+            .map(rc -> new ArrayList<>(rc.getPrepares()))
+            .orElse(new ArrayList<>()));
+  }
+
   public Prepare createPrepare(final ConsensusRoundIdentifier roundIdentifier, final Hash digest) {
+
     final PreparePayload payload = new PreparePayload(roundIdentifier, digest);
+
     return new Prepare(createSignedMessage(payload));
   }
 
   public Commit createCommit(
       final ConsensusRoundIdentifier roundIdentifier,
       final Hash digest,
       final Signature commitSeal) {
+
     final CommitPayload payload = new CommitPayload(roundIdentifier, digest, commitSeal);
+
     return new Commit(createSignedMessage(payload));
   }
 
+  public RoundChange createRoundChange(
+      final ConsensusRoundIdentifier roundIdentifier,
+      final Optional<PreparedCertificate> preparedRoundData) {
+
+    final RoundChangePayload payload;
+    if (preparedRoundData.isPresent()) {
+      final int round =
+          preparedRoundData
+              .get()
+              .getPrepares()
+              .get(0)
+              .getPayload()
+              .getRoundIdentifier()
+              .getRoundNumber();
+
+      payload =
+          new RoundChangePayload(
+              roundIdentifier,
+              Optional.of(
+                  new PreparedRoundMetadata(
+                      preparedRoundData.get().preparedBlock.getHash(), BigInteger.valueOf(round))));
```

Comment:
The ConsensusRoundIdentifier needs to use BigInteger as well. But perhaps another PR for that change?

Reply:
I think another PR would be good - need to think about what that means for IBFT2 ... i.e. can we reuse this struct, or not?